EIPS/eip-4671.md

```diff
@@ -0,0 +1,468 @@
+
+    /// @notice URI to query to get the token's metadata
+    /// @param owner Address of the token's owner
+    /// @param index Index of the token
+    /// @return URI for the token
+    function tokenURI(address owner, uint256 index) external view returns (string memory);
+}
+```
+<!-- AUTO-GENERATED-CONTENT:END -->
+
+#### Delegation
+
+An interface to standardize delegation rights of token minting.
+
+<!-- AUTO-GENERATED-CONTENT:START (CODE:syntax=solidity&src=./contracts/INTTDelegate.sol) -->
+<!-- The below code snippet is automatically added from ./contracts/INTTDelegate.sol -->
+```solidity
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+interface INTTDelegate {
+    /// @notice Grant one-time minting right to `operator` for `owner`
+    /// An allowed operator can call the function to transfer rights.
+    /// @param operator Address allowed to mint a token
+    /// @param owner Address for whom `operator` is allowed to mint a token
+    function delegate(address operator, address owner) external;
+
+    /// @notice Grant one-time minting right to a list of `operators` for a corresponding list of `owners`
+    /// An allowed operator can call the function to transfer rights.
+    /// @param operators Addresses allowed to mint
+    /// @param owners Addresses for whom `operators` are allowed to mint a token
+    function delegateBatch(address[] memory operators, address[] memory owners) external;
+
+    /// @notice Mint a token. Caller must have the right to mint for the owner.
+    /// @param owner Address for whom the token is minted
+    function mint(address owner) external;
+
+    /// @notice Mint tokens to multiple addresses. Caller must have the right to mint for all owners.
+    /// @param owners Addresses for whom the tokens are minted
+    function mintBatch(address[] memory owners) external;
+}
+```
+<!-- AUTO-GENERATED-CONTENT:END -->
+
+## Rationale
+
+### On-chain vs Off-chain
+
+A decision was made to keep the data off-chain (via `tokenURI()`) for two main reasons: 
+* Non-Tradable Tokens represent personal possessions. Therefore, there might be cases where the data should be encrypted. The standard should not outline decisions about encryption because there are just so many ways this could be done, and every possibility is specific to the use-case.
+* Non-Tradable Tokens must stay generic. There could have been a possibility to make a `MetadataStore` holding the data of NTTs in an elegant way, unfortunately we would have needed a support for generics in solidity (or struct inheritance), which is not available today.
+
+## Implementation
+
+### NTT
+
+<!-- AUTO-GENERATED-CONTENT:START (CODE:syntax=solidity&src=./contracts/NTT.sol) -->
+<!-- The below code snippet is automatically added from ./contracts/NTT.sol -->
+```solidity
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/Strings.sol";
+import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
```

Comment:
I can't keep it even if it is in the `implementation` section and not `specification` ?

Reply:
Shall I upload the implementation on a separate repo and simply reference it here ?

Reply:
I don't know if there's a "canonical" answer, but for things being published in the EIPs repo, my feeling is not to use any external libraries, though for ERCs I wouldn't mind seeing a link to an external repo (I've done that myself), though there are different opinions about this.

Reply:
All right, so I've decided to link my personal implementation rather than putting it in the EIP. That way if other implementations come up I can always put them here, and the EIP will be shorter.